extensions/amp-addthis/0.1/ConfigManager.js

```diff
@@ -0,0 +1,244 @@
+ * limitations under the License.
+ */
+
+import {isObject} from '../../../src/types';
+import {tryParseJson} from '../../../src/json';
+import {getData, listen} from '../../../src/event-helper';
+
+import {
+  WIDGET_ID_PROP,
+  CONFIGURATION_EVENT,
+  ORIGIN,
+} from './constants';
+
+/**
+ * Configuration request status enum.
+ * @enum {number}
+ */
+const RequestStatus = {
+  NOT_REQUESTED: 0,
+  REQUESTED: 1,
+  COMPLETED: 2,
+};
+
+/**
+ * Utility method to parse out the data from the supplied `postMessage` event.
+ */
+function getMessageData(event) {
+  const data = getData(event);
+
+  if (isObject(data)) {
+    return data;
+  }
+
+  if (data['indexOf']('{') === 0) {
+    return tryParseJson(data);
+  }
+
+  return;
+}
+
+/**
+ * The ConfigManager manages requests for publishers' dashboard configurations from addthis.com,
+ * ensuring that a configuration is requested for a given publisher id at most once, that
+ * all relevant amp-addthis iframes are notified when configs are received, etc. The exposed API
+ * is specified in the typedef, below; everything else is private. When an AMPElement registers
+ * with the ConfigManager, this kicks off a request for the relevant dashboard configuration, if
+ * necessary, and eventually delivers that configuration to the AMPElement's iframe. When an
+ * AMPElement unregisters, the ConfigManager destroys its references to that AMPElement's iframe and
+ * performs other cleanup when necessary (such as removing event listeners and destroying other
+ * element references).
+ */
+export class ConfigManager {
+  constructor() {
+    /**
+     * @type {!Object<string, ConfigManager.PubIdData>}
+     * @private
+     */
+    this.dataForPubId_ = {};
+
+    /**
+     * @type {!Array<Element>}
+     * @private
+     */
+    this.configProviderIframes_ = [];
+
+    /**
+     * @type {Document}
+     * @private
+     */
+    this.ownerDocument_ = null;
+
+    /**
+     * @type {function():void|null}
+     * @private
+     */
+    this.removeMessageListener_ = null;
+
+    /**
+     * Flag; only a single request should register as a page view.
+     * @type {boolean}
+     * @private
+     */
+    this.registerView_ = true;
+  }
+
+  /**
+   * Store the configuration received for the provided pubId, mark the request for the config as
+   * completed, and broadcast the config to relevant iframes.
+   * @private
+   */
+  receiveConfiguration_({config, pubId}) {
+    const pubData = this.dataForPubId_[pubId];
+    pubData.config = config;
+    pubData.requestStatus = RequestStatus.COMPLETED;
+    const {iframes} = pubData;
+    iframes.forEach(iframe => this.sendConfiguration_({iframe, pubId}));
+  }
+
+  /** @private */
+  sendConfiguration_({iframe, pubId}) {
+    const {location: loc, title} = this.ownerDocument_;
+    const pubData = this.dataForPubId_[pubId];
+    const dashboardConfig = pubData.config;
+    const configRequestStatus = pubData.requestStatus;
+    const widgetId = iframe[WIDGET_ID_PROP];
```

Comment:
It looks like this could just be passed in instead of mutating a DOM element object.